src/structarray.jl

```diff
@@ -628,3 +628,7 @@ function Broadcast.broadcast_unalias(dest::StructArray, src::AbstractArray)
 end
 
 Base.dataids(u::StructArray) = mapreduce(Base.dataids, (a, b) -> (a..., b...), values(components(u)), init=())
+
+# Since all the components have the same axes, we choose the type of the first one to
+# define IteratorSize for a StructArray
+Base.IteratorSize(::Type{<:StructArray{<:Any,<:Any,C}}) where {C} = Base.IteratorSize(fieldtype(C,1))
```

Comment:
```suggestion
Base.IteratorSize(::Type{<:StructArray{<:Any,<:Any,C}}) where {C} = Base.IteratorSize(fieldtype(C, 1))
```

There is also `array_types(S)` (where `S` is the type of a `StructArray`) to get the type of the components as a tuple type, but I guess the above is simple enough.